### PR TITLE
improve handling of GPU binaries

### PR DESCRIPTION
**Pull request by jmellorcrummey**
- add support for recording AMD OpenMP binaries
- remove unnecessary cubin-hash-map.{h,c}
- factor out common code into gpu-binary.c
- use new primitives after refactoring to reduce replication in handling for ROCM, CUDA, OpenCL, and GTPin handling of binaries

### CODE REVIEW

File: src/tool/hpcrun/gpu/instrumentation/gtpin-instrumentation.c [hunk 1]
```diff
@@ -284,20 +277,8 @@ findOrAddKernelModule
...
-    module_id = hpcrun_loadModule_add(path);
-    load_module_t *lm = hpcrun_loadmap_findById(module_id);
-    hpcrun_loadModule_flags_set(lm, LOADMAP_ENTRY_ANALYZE);
-  } else {
-    // Find module
-    module_id = module->id;
-  }
-  hpcrun_loadmap_unlock();
-  uint64_t kernel_name_id = get_numeric_hash_id_for_string(kernel_name, (size_t)kernel_name_len);
+  bool mark_used = true;
```
Comment (blue42u) [inline src/tool/hpcrun/gpu/instrumentation/gtpin-instrumentation.c:280]: This constant does not need to be a variable. Using `true /* mark_used */` in the argument list is equivalently readable. (At the very least make it `const`).

File: src/tool/hpcrun/gpu/level0/level0-command-process.c [hunk 2]
```diff
@@ -216,18 +216,9 @@ get_load_module
...
-  if (module == NULL) {
-    module_id = hpcrun_loadModule_add(load_module_name);
-    load_module_t *lm = hpcrun_loadmap_findById(module_id);
-    hpcrun_loadModule_flags_set(lm, LOADMAP_ENTRY_ANALYZE);
-  } else {
-    // Find module
-    module_id = module->id;
-  }
-  hpcrun_loadmap_unlock();
+  bool mark_used = true;
```
Comment (blue42u) [inline src/tool/hpcrun/gpu/level0/level0-command-process.c:219]: (Same as previous comment).
This constant does not need to be a variable. Using `true /* mark_used */` in the argument list is equivalently readable. (At the very least make it `const`).

File: src/tool/hpcrun/gpu/nvidia/cupti-api.c [hunk 3]
```diff
@@ -722,53 +717,17 @@ cupti_load_callback_cuda
...
-  // Create file name
-  char file_name[PATH_MAX];
-  char hash_string[PATH_MAX];
-  size_t used = 0;
-  size_t i;
-  for (i = 0; i < hash_len; ++i) {
-    used += sprintf(&hash_string[used], "%02x", hash[i]);
-  }
+  uint32_t loadmap_module_id;
+  bool mark_used = false;
```
Comment (blue42u) [inline src/tool/hpcrun/gpu/nvidia/cupti-api.c:721]: (Same as previous comment).
This constant does not need to be a variable. Using `true /* mark_used */` in the argument list is equivalently readable. (At the very least make it `const`).

File: src/tool/hpcrun/gpu/opencl/opencl-api.c [hunk 4]
```diff
@@ -496,28 +496,9 @@ opencl_write_debug_binary
...
-    if (module == NULL) {
-      uint16_t module_id = hpcrun_loadModule_add(path);
-      load_module_t *lm = hpcrun_loadmap_findById(module_id);
-      hpcrun_loadModule_flags_set(lm, LOADMAP_ENTRY_ANALYZE);
-    } 
-    hpcrun_loadmap_unlock();
-
-    free(hash_buf);
+    uint32_t loadmap_module_id;
+    bool mark_used = true;
```
Comment (blue42u) [inline src/tool/hpcrun/gpu/opencl/opencl-api.c:500]: (Same as previous comment).
This constant does not need to be a variable. Using `true /* mark_used */` in the argument list is equivalently readable. (At the very least make it `const`).

File: src/tool/hpcrun/ompt/ompt-device.c [hunk 5]
```diff
@@ -537,9 +539,9 @@ ompt_device_load(int device_num,
 {
   PRINT("ompt_device_load->%s, %d\n", filename, device_num);
 
-#if 0 // FIXME
-  cupti_load_callback_cuda(module_id, host_addr, bytes);
-#endif
+  uint32_t loadmap_module_id;
+  bool mark_used = true;
```
Comment (blue42u) [inline src/tool/hpcrun/ompt/ompt-device.c:543]: (Same as previous comment).
This constant does not need to be a variable. Using `true /* mark_used */` in the argument list is equivalently readable. (At the very least make it `const`).

File: src/tool/hpcrun/gpu/gpu-binary.c [hunk 6]
```diff
@@ -131,3 +151,56 @@ gpu_binary_compute_hash_string
...
+gpu_binary_save
+(
+ const char *mem_ptr,
+ size_t mem_size,
+ bool mark_used,
+ uint32_t *loadmap_module_id
+)
+{
+  // Generate a hash for the binary
+  char *hash_buf = (char*)malloc(HASH_LENGTH * 2);
```
Comment (blue42u) [inline src/tool/hpcrun/gpu/gpu-binary.c:191]: `hash_buf` is allocated but never freed (AFAICT).

**Review by blue42u — APPROVED** [review]
Overall LGTM, just a few style and memory leak comments.

File: src/tool/hpcrun/gpu/instrumentation/gtpin-instrumentation.c (hunk 1 above)
Comment (jmellorcrummey) [reply src/tool/hpcrun/gpu/instrumentation/gtpin-instrumentation.c]: done

File: src/tool/hpcrun/gpu/opencl/opencl-api.c (hunk 4 above)
Comment (jmellorcrummey) [reply src/tool/hpcrun/gpu/opencl/opencl-api.c]: done

File: src/tool/hpcrun/gpu/gpu-binary.c (hunk 6 above)
Comment (jmellorcrummey) [reply src/tool/hpcrun/gpu/gpu-binary.c]: replaced with stack variable to avoid leak.

File: src/tool/hpcrun/ompt/ompt-device.c (hunk 5 above)
Comment (jmellorcrummey) [reply src/tool/hpcrun/ompt/ompt-device.c]: done

File: src/tool/hpcrun/gpu/level0/level0-command-process.c (hunk 2 above)
Comment (jmellorcrummey) [reply src/tool/hpcrun/gpu/level0/level0-command-process.c]: done.

File: src/tool/hpcrun/gpu/nvidia/cupti-api.c (hunk 3 above)
Comment (jmellorcrummey) [reply src/tool/hpcrun/gpu/nvidia/cupti-api.c]: done.

**Pull request merged**